src/blight/actions/embed_commands.py

```diff
@@ -0,0 +1,125 @@
+"""
+The `EmbedCommands` action.
+
+`EmbedCommands` embeds JSON compile commands, including environment variables,
+inside of a custom section of each built object file. These sections make it
+into the final binary.
+"""
+
+import hashlib
+import json
+import os
+import random
+import shutil
+from pathlib import Path
+from typing import Dict
+
+from blight.action import CompilerAction
+from blight.enums import Lang
+from blight.tool import Tool
+from blight.util import flock_append
+
+
+def SHA256(value):
+    """Get sha256 hash from the value"""
+    return hashlib.sha256(value.encode("utf-8")).hexdigest()
+
+
+def cc_as_string(tool_record: Dict):
+    return json.dumps(tool_record).replace('"', '\\"').replace("\\\\", "\\")
+
+
+def add_to_envp(envp: Dict, key: str, value):
+    if isinstance(value, str):
+        envp[key] = value.replace('"', "'")
+    elif isinstance(value, list):
+        envp[key] = [v.replace('"', "'") for v in value]
+    else:
+        envp[key] = value
+
+
+def is_input_assembly(tool: Tool):
+    for file_path in tool.inputs:
+        file_path_str = str(file_path)
+        if file_path_str.endswith(".S") or file_path_str.endswith(".s"):
+            return True
+    return False
+
+
+def cc_as_dict(tool: Tool) -> Dict:
+    env = {}
+    tool_dict = tool.asdict()
+    old_env = tool_dict["env"]
+    for key, value in old_env.items():
+        if key == "PS1" or key == "LS_COLORS" or key.startswith("BLIGHT"):
+            continue
+        add_to_envp(env, key, value)
+
+    for key, value in old_env.items():
+        if key.startswith("BLIGHT_WRAPPED_"):
+            add_to_envp(env, key[15:], value)
+
+    return {
+        "cwd": tool_dict["cwd"],
+        "env": env,
+        "args": [v.replace('"', "'") for v in tool.args],
+        "canonicalized_args": [v.replace('"', "'") for v in tool.canonicalized_args],
+        "wrapped_tool": shutil.which(tool.wrapped_tool()),
+        "lang": tool.lang.name,
+    }
+
+
+_ALIGN = 4
+
+
+def align_string(string: str) -> str:
+    length = len(string)
+    length_aligned = length if length % _ALIGN == 0 else (int(length / _ALIGN) + 1) * _ALIGN
+    return string.ljust(length_aligned, "\0")
+
+
+class EmbedCommands(CompilerAction):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def _get_header_file(self, cmd_hash: str):
+        output = Path(self._config["output"])
+        header_file = "{}/{}_{}_{}.h".format(
+            output, cmd_hash, os.getpid(), random.randint(0, 9999999)
```

Comment:
Question: What's up with `random.randint(0, 9999999)`? If this is about providing a non-colliding header file, maybe we can use `mktemp` or similar?